src/sources/amqp.rs

```diff
@@ -0,0 +1,315 @@
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(deny_unknown_fields)]
+pub struct AmqpSourceConfig {
+    pub(crate) queue: String,
+    pub(crate) consumer: String,
+    pub(crate) routing_key: Option<String>,
+    pub(crate) exchange_key: Option<String>,
+    pub(crate) offset_key: Option<String>,
+    pub(crate) connection: AmqpConfig,
+}
+
+impl Default for AmqpSourceConfig {
+    fn default() -> Self {
+        Self {
+            queue: "vector".to_string(),
+            consumer: "vector".to_string(),
+            routing_key: None,
+            exchange_key: None,
+            offset_key: None,
+            connection: AmqpConfig::default(),
+        }
+    }
+}
+
+inventory::submit! {
+    SourceDescription::new::<AmqpSourceConfig>("amqp")
+}
+
+impl_generate_config_from_default!(AmqpSourceConfig);
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "amqp")]
+impl SourceConfig for AmqpSourceConfig {
+    async fn build(
+        &self,
+        cx: SourceContext,
+    ) -> crate::Result<super::Source> {
+        amqp_source(self, cx.shutdown, cx.out).await
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn source_type(&self) -> &'static str {
+        "amqp"
+    }
+}
+
+enum ShutdownOrMessage {
+    Shutdown,
+    Message(Option<Result<(lapin::Channel, lapin::message::Delivery), lapin::Error>>),
+}
+
+async fn amqp_source(
+    config: &AmqpSourceConfig,
+    shutdown: ShutdownSignal,
+    out: Pipeline,
+) -> crate::Result<super::Source> {
+    let config = config.clone();
+    let (conn, channel) = config.connection.connect().await.map_err(|e| BuildError::AmqpCreateError {
+        source: e,
+    })?;
+
+    Ok(Box::pin(run_amqp_source(config, shutdown, out, conn, channel)))
+}
+
+async fn run_amqp_source(config: AmqpSourceConfig, shutdown: ShutdownSignal, mut out: Pipeline, _conn: Connection, channel: Channel) -> Result<(), ()> {
+    let ack_options = lapin::options::BasicAckOptions::default();
+    debug!("Starting amqp source, listening to queue {}", config.queue);
+    let mut consumer = channel
+        .basic_consume(
+            &config.queue,
+            &config.consumer,
+            lapin::options::BasicConsumeOptions::default(),
+            lapin::types::FieldTable::default(),
+        )
+        .await.map_err(|error| {
+        emit!(AmqpConsumerFailed { error });
+        ()
+    })?.fuse();
+    let mut shutdown = shutdown.fuse();
+    loop {
+        let msg = futures::select! {
+            _ = shutdown => ShutdownOrMessage::Shutdown,
+            opt_m = consumer.next() => ShutdownOrMessage::Message(opt_m),
+        };
+        if let ShutdownOrMessage::Message(Some(try_m)) = msg {
+            match try_m {
+                Err(error) => {
+                    emit!(AmqpEventFailed { error });
+                    return Err(());
+                }
+                Ok((_, msg)) => {
+                    debug!("Amqp event received");
```

Comment:
I think we can drop this trace message, it should be output by the internal events.